Plex Media Server

Forkk opened the pull request:
Added a package and module for Plex Media Server, an application for managing media collections across multiple devices.

This is a bit of a tricky application, as it stores some databases and other mutable files alongside some binaries and other immutable files in a directory under `/usr/lib/plexmediaserver/Resources`.

My current solution to this problem is to rename the `Resources` folder to `base-resources` on install and replace the original `Resources` folder with a symlink to `/var/lib/plex/Resources`. Then, in the systemd service, I set the `ExecStartPre` command to copy the files from `base-resources` to `/var/lib/Resources` if the latter directory is not present. This fixes the issue and, thus, Plex works properly on NixOS.

The potential issue here is that if, in the future, the some of the executables in the `Resources` directory need to be updated, we will have no suitable mechanism in place to do so. One option is to keep the immutable files in the Nix store and only keep the mutable ones in `/var/`, but unfortunately, I don't know of a very reliable way to determine which files are mutable and which aren't.

I would greatly appreciate feedback on this issue, as I'm a bit new here and I feel like there could be something I am missing that might solve this problem. Thanks. :)

Forkk committed:
plex: init at 0.9.11.16.958

Added a package and module for Plex Media Server, an application for
managing media collections across multiple devices.

nbp commented on nixos/modules/services/misc/plex.nix:
```diff
@@ -0,0 +1,90 @@
+      };
+    };
+
+    users.extraUsers = optionalAttrs (cfg.user == "plex") (singleton {
```
use `mkIf` instead of `optionalAttrs`.

nbp commented on nixos/modules/services/misc/plex.nix:
```diff
@@ -0,0 +1,90 @@
+      serviceConfig = {
+        Type = "simple";
+        User = "plex";
+        Group = "plex";
```
I tihnk you meant `cfg.user`, and `cfg.group`.

edolstra replied:
Actually, why does the user/group need to be configurable at all?

Forkk replied:
It really doesn't, but it takes little extra effort to make it configurable, so I don't see any reason not to.

jagajaga commented on nixos/modules/services/misc/plex.nix:
```diff
@@ -0,0 +1,90 @@
+{
+  options = {
+    services.plex = {
+      enable = mkOption {
```
Use `mkEnableOption`.

jagajaga commented on nixos/modules/services/misc/plex.nix:
```diff
@@ -0,0 +1,90 @@
+        Type = "simple";
+        User = "plex";
+        Group = "plex";
+        #Restart = "on-failure";
```
Remove these commented lines if we don't need them?

Forkk commented:
Do not merge yet. I'm waiting to hear back from the Plex guys on whether or not the `Resources` directory is meant to be read-only. (As mentioned in [this tweet](https://twitter.com/plexapp/status/588779473635409920).)

If this is the case (and I really hope it is), then the `Resources` dir problem doesn't exist.

Forkk commented:
Haven't heard back from the Plex people, but I did run a diff on the base resources directory and the one that was copied to `/var` and it looks like none of the files in are being changed (aside from .pyc files being added due to Python).

According to what they've said on Twitter, Plex only uses the files in the `Resources` directory as base files, which it copies to its directory in `/var/lib`. If this is true (as it seems to be), then the `Resources` directory *should* be included with the read-only files in the Nix store, however, the issue with this is that when Plex starts up, it attempts to open some SQLite databases in the `Resources` directory (although it evidently doesn't modify them). Since the directory is read-only, SQLite chokes and spits out an error when opening the database files, causing Plex to fail to start. I guess a solution to this would be to just put the database file in a mutable location and copy it there on every startup. This still seems a bit too hacky, though.

Forkk commented:
Alright, I've fixed the issue.

The `Resources` directory remains read-only. The database files are moved to base_* and replaced with symlinks to `/var/lib/plex/.skeleton/*`. On startup, the systemd service just throws them all into that folder and Plex doesn't crash due to read-only databases.

At this point, I would consider this ready for merging.

jagajaga commented:
Please rebase against current master. There are conflicts :(

Forkk commented:
Merge conflicts have been fixed. :)

anderslundstedt commented:
After enabling plex

    nixpkgs.config.allowUnfree = true;
    services.plex.enable = true;

and doing `nixos-rebuild switch` I got

    Jun 09 02:30:40 nixos systemd[1]: Starting Plex Media Server...
    Jun 09 02:30:40 nixos plex-pre-start[28055]: Creating initial Plex data directory in "/var/lib/plex".
    Jun 09 02:30:40 nixos plex-pre-start[28055]: Preparing plugin directory.
    Jun 09 02:30:40 nixos plex-pre-start[28055]: Removing old symlinks.
    Jun 09 02:30:40 nixos plex-pre-start[28055]: Symlinking plugins.
    Jun 09 02:30:40 nixos systemd[1]: Started Plex Media Server.
    Jun 09 02:30:40 nixos sh[28069]: terminate called after throwing an instance of 'boost::filesystem::filesystem_error'
    Jun 09 02:30:40 nixos sh[28069]: what():  boost::filesystem::create_directory: Permission denied: "/var/lib/plex/Plex Media Server/Logs"
    Jun 09 02:30:40 nixos sh[28069]: ****** PLEX MEDIA SERVER CRASHED, CRASH REPORT WRITTEN: /var/lib/plex/Plex Media Server/Crash Reports/1ab8e852-c01f-24bb-3c94f5d1-63497a84-v-0.9.11.16.958-80f1748.dmp
    Jun 09 02:30:40 nixos systemd[1]: plex.service: main process exited, code=killed, status=6/ABRT
    Jun 09 02:30:40 nixos systemd[1]: Unit plex.service entered failed state.
    Jun 09 02:30:40 nixos systemd[1]: plex.service failed.

After

    chown plex:plex /var/lib/plex/.skeleton
    chown plex:plex /var/lib/plex/Plex\ Media\ Server
    systemctl restart plex

everything works great!
